File: src/main/kotlin/ex09/java/Exam09.java
package ex09.java;

public class Exam09 {

    public Exam09(StoreUser storeUser){
        // service logic

        // if문 길어지면 가독성 떨어짐
        if (storeUser.getRole().equals("MASTER")){

        } else if (storeUser.getRole().equals("ADMIN")) {

        } else if (storeUser.getRole().equals("USER")) {

        }else {

        }

        // 그렇다고 switch문을 사용하면 break를 실수로 안적을 수도 있다
        var role = "";
        switch (storeUser.getRole()){
            case "MASTER":
                role = "MASTER"; //break를 실수로 안적을 수도 있다
            case "ADMIN":
                role = "ADMIN";
                break;
            case "USER":
            default:
                break;
        }

        try {

        }catch (Exception e){
            if(e instanceof NullPointerException){

            } else if (e instanceof  NumberFormatException) {
                // 이렇게 해도 e의 타입이 Exception라 (NumberFormatException)로 캐스트 해줘야함
            }
        }
    }

    public static void main(String[] args){

    }
}

class StoreUser{
    private String name;
    private String role;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getRole() {
        return role;
    }

    public void setRole(String role) {
        this.role = role;
    }
}

File: src/main/kotlin/ex09/Exam09.kt
package ex09

import java.lang.NumberFormatException
import java.lang.RuntimeException
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

fun main(){
    val role: String= "MASTER"
    // break가 없어 실수할 가능성도 없고 가독성도 좋음
    val result = when(role){
        "MASTER", "ADMIN" -> {

        }
        "USER" -> {

        }
        else -> {

        }
    }

    // 타입 체크
    var any: Any = "";
    when(any){
        is String -> { // instance of
            any.length // 이미 스코프안에서 타입 캐스팅이 일어남
        }
        is Int -> {
            any.compareTo(10L)
        }
    }
    
    var exception = RuntimeException();
    when(exception){
        is NumberFormatException ->{
            // 스코프안 타입캐스팅
        }
        is RuntimeException ->{
        }
    }

    // 표현식도 가능
    var number = 10
    when(val n = number % 2){ // 표현식도 가능
        0 -> {
            println("$n 은 짝수")
        }
        else -> {
            println("$n 은 홀수")
        }
    }
    
    val r = when{
        number % 2 == 0 ->{
            "짝수"
        } 
        else -> {
            "홀수"
        }
    }
}
